src/V3Delayed.cpp

```diff
@@ -118,307 +125,297 @@ class DelayedVisitor final : public VNVisitor {
-        if (bitselp) {
-            AstNodeExpr* const lsbvaluep = bitselp->lsbp()->unlinkFrBack();
-            if (VN_IS(lsbvaluep, Const)) {
-                // vlsb = constant, can just push constant into where we use it
-                bitreadp = lsbvaluep;
-            } else {
-                const string bitvarname
-                    = string{"__Vdlyvlsb__"} + oldvarp->shortName() + "__v" + cvtToStr(modVecNum);
-                AstVarScope* const bitvscp
-                    = createVarSc(varrefp->varScopep(), bitvarname, lsbvaluep->width(), nullptr);
-                AstAssign* const bitassignp = new AstAssign{
-                    nodep->fileline(), new AstVarRef{nodep->fileline(), bitvscp, VAccess::WRITE},
-                    lsbvaluep};
-                nodep->addNextHere(bitassignp);
-                bitreadp = new AstVarRef{nodep->fileline(), bitvscp, VAccess::READ};
-            }
+        // Done
+        return nodep;
+    }
+
+    void createDlyOnSet(AstAssignDly* nodep) {
+        // Create delayed assignment
+        // See top of this file for transformation
+        // Return the new LHS for the assignment, Null = unlink
+
+        // Insertion point/helper for adding new statements in code order
+        AstNode* insertionPointp = nodep;
+        const auto insert = [&insertionPointp](AstNode* stmtp) {
+            insertionPointp->addNextHere(stmtp);
+            insertionPointp = stmtp;
+        };
+
+        // Deconstruct the LHS
+        LhsComponents lhsComponents = deconstructLhs(nodep->lhsp()->unlinkFrBack());
+
+        // The referenced variable
+        AstVarScope* const vscp = lhsComponents.refp->varScopep();
+
+        // Name suffix for signals constructed below
+        const std::string baseName
+            = "__" + vscp->varp()->shortName() + "__v" + std::to_string(m_scopeVecMap[vscp]++);
+
+        // Given an 'expression', return a new expression that always evaluates to the value of the
+        // given expression at this point in the program. That is:
+        // - If given a non-constant expression, create a new temporary AstVarScope with the given
+        //   'name' prefix, assign the expression to it, and return a read reference to the new
+        //   AstVarScope.
+        // - If given a constant, just return that constant.
+        const auto capture = [&](AstNodeExpr* exprp, const std::string& name) -> AstNodeExpr* {
+            UASSERT_OBJ(!exprp->backp(), exprp, "Should have been unlinked");
+            if (VN_IS(exprp, Const)) return exprp;
+            const std::string realName = "__" + name + baseName;
+            AstVarScope* const tmpVscp = createNewVarScope(vscp, realName, exprp->dtypep());
+            FileLine* const flp = exprp->fileline();
+            insert(new AstAssign{flp, new AstVarRef{flp, tmpVscp, VAccess::WRITE}, exprp});
+            return new AstVarRef{flp, tmpVscp, VAccess::READ};
+        };
```

Comment:
Well, any closure could be by passing everything explicitly, but this captures `baseName`, `vscp` (and in one of the later patched a different `flp`), so It would be clearer not to. With the lambda it's clear what is common, what is variable.